src/wh_server_dma.c

```diff
@@ -0,0 +1,307 @@
+#include "wolfhsm/wh_server.h"
+#include "wolfhsm/wh_error.h"
+#include "wolfhsm/wh_server_dma.h"
+
+/* TODO: if the Address allowlist ever gets large, we should consider a more
+ * efficient representation (requiring sorted array and binary search, or
+ * building a binary tree, etc.) */
+
+
+static int _checkAddrAgainstAllowList(const whServerDmaAddrList allowList, void* addr,
+                                      size_t size)
+{
+    uintptr_t startAddr = (uintptr_t)addr;
+    uintptr_t endAddr   = startAddr + size;
+
+    /* Check if the address range is fully within a allowlist entry */
+    for (int i = 0; i < WH_DMA_ADDR_ALLOWLIST_SIZE; i++) {
+        uintptr_t allowlistStartAddr = (uintptr_t)allowList[i].addr;
+        uintptr_t allowlistEndAddr   = allowlistStartAddr + allowList[i].size;
+
+        if (startAddr >= allowlistStartAddr && endAddr <= allowlistEndAddr) {
```

Comment:
If allowList[].size == 0, then you shouldn't check for an allow.  Currently, a size==0 will allow addr==allowList[].addr.  I  also think an incoming size of 0 should always fail with WH_ERROR_BADARGS